### DISCUSSION TITLE
What is the easiest or most useful target to build?

### DISCUSSION THREAD

**Discussion by squinkylabs** [sfztools/sfizz#608] [Q&A]
Looking for advice. I want to build some form of sfizz so that I can run it in a debugger and learn more about the architecture. My near term goal is to make a command-line windows app that I can “goof around with”.
I’m hoping/assuming that I can get cmake to make something for me, although I have also tried making my own Visual Studio project for it.
I am guessing that turning off various targets will increate my chance of success. So here is my current invocation of CMake. It seems like I’m close, but if you have any suggestions it would be appreciated.

cmake -DCMAKE_BUILD_TYPE=Debug -DSFIZZ_JACK=OFF -DSFIZZ_SHARED=OFF -DSFIZZ_RENDER=OFF ..

**Comment by jpcima** [comment]
To not use cmake at this moment is to ask for more difficulty, it's much recommendable to stick to it.
This takes care automatically of processing the external dependencies.

Something else, `RelWithDebInfo` is a better configuration for developing than `Debug`, as unoptimized builds are next to unusable.

Besides that, it looks pretty OK. Really we should edit the build system such that Jack is not enabled by default in Windows and OSX platforms, I'm going to do this shortly.

**Comment by squinkylabs** [comment]
Yeah, I actually got it to build! thanks. looking at my generated "stuff", it certainly looks like lv2 was built (and some other stuff). w00t!

Thinking about how to make a hacky "test command line app" - one way that is sitting in front of my nose is to clone the "sfizz_lv2" project to make a "sfizz_myhack" project, import it back into the big "sfizz.sln" and then hack it up until it makes my code instead of making lv2.  

I'm pretty sure that would eventually work, but maybe there is any easier way?

**Reply by jpcima** [reply]
The indicated way to do this, it's to have a clone of sfizz as subdirectory of your project, and create a cmake build file that includes sfizz, as given in a previous example.

From that point on, cmake acts as a generator of the Visual Studio project, which means you needn't touch any .sln files; this becomes under cmake's management. You're just supposed to list your sources, your libraries in text form in the cmake file.

The more precise way that you want to handle it, it depends what type of app you have in mind.
One solution among others that you can do, is to build software on the Juce libraries, and you get midi and audio handled for you, that you can then interact to sfizz.

At some point if you or someone wishes it, I can provide a template project with sfizz based on Juce6.

Hope this helps

**Reply by squinkylabs** [reply]
so in your example:
```cmake_minimum_required(VERSION 3.7)

add_subdirectory(external/sfizz EXCLUDE_FROM_ALL)

add_executable(myprogram PRIVATE sources/myprogram.cpp)
target_link_libraries(myprogram PRIVATE sfizz::sfizz)```

I would literally add all of sfizz with a single line of cmake? the add_subdirectory will cause cmake to find everything it needs to build sfizz?  Or is the idea that you FIRST run cmake on sfizz? Sorry for these dumb questions, but I've never dealt with something like this that creates nested VS projects. It all seems like black magic to me ;-)

fwiw - I was able to easily make a command line program to talk to sfz::Synth by hacking up the sfizz_lv2 project.

**Reply by jpcima** [reply]
> I would literally add all of sfizz with a single line of cmake?

Precisely, that's it. You run cmake only from your top folder.
The "exclude from all" setting tells it that it must build only these targets that your project needs, directly or indirectly.
Hence, this will build the sfizz library, but ignore any plugins.

**Reply by jpcima** [reply]
It's no longer forced to set JACK to OFF in latest dev, it's automatic.
